air/src/execution_step/instructions/fold/utils.rs

```diff
@@ -104,10 +106,27 @@ pub(crate) fn create_canon_stream_map_iterable_value(
         return Ok(FoldIterableScalar::Empty);
     }
 
-    // TODO: this one is a relatively long operation and will be refactored in Boxed Value
+    // TODO: this one is a relatively heavy operation and will be refactored in Boxed Value
     // Can not create iterable from existing CanonStreamMap b/c CSM contains a map with
     // a limited lifetime but the boxed value needs static lifetime.
-    let values = canon_stream_map.iter().cloned().collect::<Vec<_>>();
+    let mut met_keys = HashSet::new();
+    let values = canon_stream_map
+        .iter()
+        .rev()
+        .filter(|&val| {
+            if let Some(map_key) = StreamMapKey::from_kvpair_owned(val) {
+                if met_keys.get(&map_key).is_some() {
+                    false
+                } else {
+                    met_keys.insert(map_key);
+                    true
+                }
+            } else {
+                false
```

Comment:
I think complying with filter interface is too much hassle here.
Try to write it with a simple for loop (pushing to a Vec) to see if it sorter and clearer.

Reply:
Replaced. The suggested approach delivers a smaller nibble for sure.